### PR TITLE
Keep track of own Usermask (nickmask)

### PR DESCRIPTION
**Pull request by ThePhilgrim**
This is a slightly hacky way to keep track of the usermask (which is actually called nickmask according to the Horse docs). It relies on sending a `WHOIS own_nick` to the server which then always replies with all components of the nickmask in `311 RPL_WHOISUSER`.

Ex:
```
Sent – WHOIS fubbe
Received – :copper.libera.chat 311 Fubbe Fubbe ~IAmTheFub 151.40.115.120 * :lol <---------
Received – :copper.libera.chat 312 Fubbe Fubbe copper.libera.chat :Sofia, BG
Received – :copper.libera.chat 338 Fubbe Fubbe 151.40.115.120 :actually using host
Received – :copper.libera.chat 317 Fubbe Fubbe 6 1644854035 :seconds idle, signon time
Received – :copper.libera.chat 318 Fubbe fubbe :End of /WHOIS list.
```

This should help to make it easier solving #172 (in a separate PR). This code would have to check if `server_view.core.nickmask != ""` to make sure the messages are not cut off incorrectly. However, most respectable IRC servers support `WHOIS`

Ex Freenode:
```
WHOIS fubbe
:teepee.freenode.net 311 fubbe fubbe ~bubbe joseon-8tr.tfj.n3gjid.IP * :lol
```

Of course, this logic can be substituted for something like `USERLEN` if Libera ever implements it, or if we find something else that would be a more elegant solution.

Right now, the test xfails. This is because neither Mantatail nor Hircd support `WHOIS`. This is noted in https://github.com/ThePhilgrim/MantaTail/issues/128

If you feel that this is not a good enough solution, feel free to close.

### CODE REVIEW

**Comment by ThePhilgrim** [comment]
The `if` check in [66891f5](https://github.com/Akuli/mantaray/pull/244/commits/66891f5d9bbc982740aa658c9b539bb3ab6b3795) ensures that manual use of `WHOIS` can be implemented (#97)

**Comment by ThePhilgrim** [comment]
I don't really understand the test errors. I didn't touch anything related to the tests apart from the one I wrote myself, which shouldn't affect any other tests.

File: mantaray/received.py [hunk 1]
```diff
@@ -484,6 +488,10 @@ def _handle_received_message(
     elif msg.command == RPL_TOPIC:
         _handle_numeric_rpl_topic(server_view, msg.args)
 
+    elif msg.command == RPL_WHOISUSER:
+        if msg.args[0] == server_view.core.nick:
+            server_view.core.nickmask = f"{msg.args[1]}!{msg.args[2]}@{msg.args[3]}"
```
Comment (Akuli) [inline mantaray/received.py:493]: I would prefer to store only the `!user@host` part into the core, without the nickname. That way it's impossible to get a bug where the nick changes in either `core.nick` or `core.nickmask`, but not both. The rest of the mask will never change, unless you reconnect. We can then add a `get_user_mask()` method that returns the mask with `self.nick` in front.

File: tests/test_config.py [hunk 2]
```diff
@@ -173,3 +173,29 @@ def bob_config(transient_to, initial_config):
     assert "left" not in bob.text()
     assert "parted" not in bob.text()
     assert "quit" not in bob.text()
+
+
+# Neither Mantatail nor Hircd currently supports the WHOIS command.
+# IrcCore.nickmask will therefore always be "" in the test.
```
Comment (Akuli) [inline tests/test_config.py:179]: We could also capture it when you join a channel, which will work most of the time. The `WHOIS` thing can still be useful. For example, if Alice is angry, and wants to write a long rant to Bob in a private message, she probably doesn't want to join any public channels if she can speak to Bob directly.

File: mantaray/backend.py [hunk 3]
```diff
@@ -135,6 +135,8 @@ def __init__(self, server_config: config.ServerConfig, *, verbose: bool):
         self.event_queue: queue.Queue[IrcEvent] = queue.Queue()
         self._thread: threading.Thread | None = None
 
+        self.nickmask = ""
```
Comment (Akuli) [inline mantaray/backend.py:138]: I want mypy to warn me, if I forget that the mask might not be available:

```suggestion
        self.nickmask: str | None = None
```

**Review by Akuli — COMMENTED** [review]
I like the idea of using WHOIS :)

File: tests/test_config.py (hunk 2 above)
Comment (ThePhilgrim) [reply tests/test_config.py:179]: I'm not completely sure what the point of this comment is. It was 9 days ago to be fair :)
What exactly is it you want to capture when joining a channel?

**Comment by ThePhilgrim** [comment]
Working on `WHOIS` in Mantatail to make sure the tests pass.

File: tests/test_config.py (hunk 2 above)
Comment (Akuli) [reply tests/test_config.py:228]: When you join a channel, the IRC server sends you the user mask that we can then use, even if the server doesn't support WHOIS.

File: mantaray/backend.py [hunk 4]
```diff
@@ -199,6 +201,10 @@ def get_events(self) -> list[IrcEvent]:
         self._events.clear()
         return result
 
+    def get_nickmask(self) -> str:
+        assert self.nickmask is not None
```
Comment (Akuli) [inline mantaray/backend.py:205]: If the nickmask isn't available, it shouldn't be an error, so that hircd is still usable. Any code using the nickmask must be prepared to handle it being missing.

```suggestion
    def get_nickmask(self) -> str | None:
        if self.nickmask is None:
            return None
```

File: mantaray/backend.py [hunk 5]
```diff
@@ -185,6 +185,8 @@ def __init__(self, server_config: config.ServerConfig, *, verbose: bool):
         self._ping_sent = False
         self._last_receive_time = time.monotonic()
 
+        self.nickmask: str | None = None
```
Comment (Akuli) [inline mantaray/backend.py:188]: - Please name this starting with `_` to prevent me from accidentally accessing it outside this class.
- This should be set to `None` when reconnecting. Compare it to `_ping_sent`, for example.

File: mantaray/received.py [hunk 6]
```diff
@@ -208,6 +209,9 @@ def _handle_nick(server_view: views.ServerView, old_nick: str, args: list[str])
             )
             if isinstance(view, views.ChannelView):
                 view.userlist.change_nick(old_nick, new_nick)
+
+        server_view.core.send(f"WHOIS {new_nick}")
```
Comment (Akuli) [inline mantaray/received.py:213]: It isn't necessary to have this in `_handle_nick()`, because `get_nickmask()` uses `core.nick` which was updated above.

**Comment by ThePhilgrim** [comment]
@Akuli I believe this is ready for another review

File: mantaray/received.py [hunk 7]
```diff
@@ -590,6 +594,9 @@ def _handle_received_message(
     elif msg.command == RPL_TOPIC:
         _handle_numeric_rpl_topic(server_view, msg.args)
 
+    elif msg.command == RPL_WHOISUSER:
+        if msg.args[0] == server_view.core.nick:
+            server_view.core._nickmask = f"!{msg.args[2]}@{msg.args[3]}"
```
Comment (Akuli) [inline mantaray/received.py:599]: Naming an attribute `_nickmask` convention is supposed to mean that it is accessed only inside a single class, not outside it. Can you add a `set_nickmask()` method to the core that works like this?

```suggestion
            server_view.core.set_nickmask(user=msg.args[2], host=msg.args[3])
```

File: tests/test_config.py [hunk 8]
```diff
@@ -224,6 +224,32 @@ def bob_config(transient_to, initial_config):
     assert "quit" not in bob.text()
 
 
+# Hircd currently doesn't support the WHOIS command
+# IrcCore._nickmask will therefore always be None in the test.
```
Comment (Akuli) [inline tests/test_config.py:228]: This text should go into the `reason="..."` string, which is currently wrong as it was copy/pasted.

File: tests/test_config.py [hunk 9]
```diff
@@ -224,6 +224,32 @@ def bob_config(transient_to, initial_config):
...
+@pytest.mark.skipif(
+    os.environ["IRC_SERVER"] == "hircd", reason="hircd sends QUIT twice"
+)
+@pytest.mark.xfail
+def test_generate_nickmask(alice, mocker, wait_until):
+    server_view = alice.get_server_views()[0]
+    assert alice.server_view.core._nickmask == "Alice!AliceUsr@localhost"
```
Comment (Akuli) [inline tests/test_config.py:235]: Presumably these should call `get_nickmask()`.

File: tests/test_config.py [hunk 10]
```diff
@@ -224,6 +224,32 @@ def bob_config(transient_to, initial_config):
     assert "quit" not in bob.text()
 
 
+# Hircd currently doesn't support the WHOIS command
+# IrcCore._nickmask will therefore always be None in the test.
+@pytest.mark.skipif(
+    os.environ["IRC_SERVER"] == "hircd", reason="hircd sends QUIT twice"
+)
+@pytest.mark.xfail
```
Comment (Akuli) [inline tests/test_config.py:232]: If I understand correctly, this no longer needs to be xfailed on mantaray.

File: mantaray/backend.py [hunk 11]
```diff
@@ -245,6 +255,7 @@ def run_one_step(self) -> None:
 
             self._ping_sent = False
             self._last_receive_time = time.monotonic()
+            self._nickmask = None
```
Comment (Akuli) [inline mantaray/backend.py:258]: Why is this here, instead of under the "Clearing data from previous connections" comment?

File: tests/test_config.py [hunk 12]
```diff
@@ -224,6 +224,28 @@ def bob_config(transient_to, initial_config):
     assert "quit" not in bob.text()
 
 
+@pytest.mark.skipif(
+    os.environ["IRC_SERVER"] == "hircd",
+    reason="Hircd currently doesn't support the WHOIS command. IrcCore._nickmask will therefore always be None in the test.",
```
Comment (Akuli) [inline tests/test_config.py:229]: If it's always `None`, you might as well assert it in the test. For example:
```python3
if os.environ["IRC_SERVER"] == "hircd":
    assert server_view.core.get_nickmask() is None
else:
    assert server_view.core.get_nickmask() == Foo!AliceUsr@127.0.0.1"
```

This ensures that `get_nickmask()` works even when the server is somewhat dumb.

**Review by Akuli — APPROVED** [review]
